src/libaktualizr/primary/aktualizr.h

```diff
@@ -13,6 +13,38 @@
 #include "uptane/secondaryinterface.h"
 #include "utilities/events.h"
 
+class ApiQueue {
+ public:
+  void enqueue(std::function<void()> t) {
+    std::lock_guard<std::mutex> lock(m_);
+    q_.push(t);
+    c_.notify_one();
+  }
+
+  std::function<void()> dequeue(void) {
+    std::unique_lock<std::mutex> wait_lock(m_);
+    while (q_.empty()) {
+      c_.wait(wait_lock);
+      if (shutdown_) {
+        return std::function<void()>();
+      }
+    }
+    std::function<void()> val = q_.front();
+    q_.pop();
+    return val;
+  }
+  void shutDown() {
+    shutdown_ = true;
+    c_.notify_all();
+  }
+
+ private:
+  std::queue<std::function<void()>> q_;
+  mutable std::mutex m_;
+  std::condition_variable c_;
+  std::atomic_bool shutdown_{false};
```

Comment:
Can this reuse the `shutdown_` member of the Aktualizr class? It seems odd to have two variables that seem to do the same thing.

Reply:
We can use only shutdown_ from queue in Aktualizr class, but I am not sure it is good idea. What do you think?

Reply:
What would be the problem with only using one? I just don't like the redundancy of having two variables that to the same thing. It seems like you basically have to set both to actually get the system to shut down. Or am I missing something? If they really are different, maybe we should rename one of them.

Reply:
No, you are probably right. It's fine how it is. It was just my overactive sense for removing redundancy!